File: src/day3.clj
(ns day3
  (:require
    [clojure.java.io :as io]
    [clojure.string :as str]))


(def sample-input
  (slurp (io/resource "day3-sample.txt")))


(def actual-input
  (slurp (io/resource "day3-actual.txt")))


(defn ^:private descend
  [rows dr dc]
  (loop [r 0 c 0 s 0]
    (if-let [v (and (< r (count rows)) (nth (nth rows r) c))]
      (let [i (if (= v \#) 1 0)]
        (recur (+ r dr) (+ c dc) (+ s i)))
      s)))


(defn solution
  [input]
  (let [lines (str/split-lines input)
        rows  (map #(cycle (seq (char-array %))) lines)]
    (->> [[1 1] [1 3] [1 5] [1 7] [2 1]]
         (map (fn [[dr dc]] (descend rows dr dc)))
         (reduce *))))


(solution sample-input)

(solution actual-input)

File: hs/src/Day3.hs
-- |
module Day3 where

type Coordinate = (Int, Int)

type Map = [String]

data Cell = Open | Tree deriving (Enum)

paths :: [Coordinate]
paths =
  [ (1, 1),
    (1, 3),
    (1, 5),
    (1, 7),
    (2, 1)
  ]

cell :: Char -> Int
cell '#' = 1
cell _ = 0

cellAt :: Map -> Coordinate -> Int
cellAt rows (r, c) = cell ((rows !! r) !! c)

search :: Map -> Coordinate -> Coordinate -> Int -> Int
search rows (dr, dc) (r, c) s
  | r < length rows =
    search rows (dr, dc) (r + dr, c + dc) (s + cellAt rows (r, c))
  | otherwise = s

descend :: Map -> Coordinate -> Int
descend rows path =
  search rows path (0, 0) 0

pathProduct :: String -> Int
pathProduct content =
  let rows = map cycle $ lines content
   in product $ map (descend rows) paths

part1 :: IO Int
part1 = pathProduct <$> readFile "../input/day3-actual.txt"
